### PR TITLE
Bitmasked multi columns

### PR DESCRIPTION
**Pull request by bancer**
resolves https://github.com/dereuromark/cakephp-tools/issues/140

### CODE REVIEW

File: Model/Behavior/BitmaskedBehavior.php [hunk 1]
```diff
@@ -37,14 +37,41 @@ class BitmaskedBehavior extends ModelBehavior {
...
+	 * ];
+	 * </code>
 	 *
 	 * @param Model $Model
 	 * @param array $config
 	 * @return void
 	 */
 	public function setup(Model $Model, $config = []) {
-		$config += $this->_defaultConfig;
+		if (is_array(reset($config))) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: whats reset() here?

File: Model/Behavior/BitmaskedBehavior.php [hunk 2]
```diff
@@ -37,14 +37,41 @@ class BitmaskedBehavior extends ModelBehavior {
...
+				$fieldName = $fieldConfig['field'];
+				$this->settings[$Model->alias][$fieldName] = $this->_getFieldConfig($Model, $fieldConfig);
+			}
+		} else {
+			$fieldName = $config['field'];
+			$this->settings[$Model->alias][$fieldName] = $this->_getFieldConfig($Model, $config);
+		}
+	}
 
+	private function _getFieldConfig(Model $Model, $config) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: protected - and doc block missing

File: Model/Behavior/BitmaskedBehavior.php [hunk 3]
```diff
@@ -85,17 +108,17 @@ public function beforeFind(Model $Model, $query) {
...
-		if (!($mappedField = $this->settings[$Model->alias]['mappedField'])) {
-			$mappedField = $field;
-		}
-
-		foreach ($results as $key => $result) {
-			if (isset($result[$Model->alias][$field])) {
-				$results[$key][$Model->alias][$mappedField] = $this->decodeBitmask($Model, $result[$Model->alias][$field]);
+		foreach ($this->settings[$Model->alias] as $fieldConfig) {
+			$field = $fieldConfig['field'];
+			if (!($mappedField = $fieldConfig['mappedField'])) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: no inline assignment please.

File: Model/Behavior/BitmaskedBehavior.php [hunk 4]
```diff
@@ -168,29 +194,30 @@ public function encodeBitmask(Model $Model, $value, $defaultValue = null) {
...
+				$mappedField = $field;
+			}
+			foreach ($conditions as $key => $val) {
+				if ($key === $mappedField) {
+					$conditions[$field] = $this->encodeBitmask($Model, $val);
+					if ($field !== $mappedField) {
+						unset($conditions[$mappedField]);
+					}
+					continue;
+				} elseif ($key === $Model->alias . '.' . $mappedField) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: no else required after continue (abort early).

File: Model/Behavior/BitmaskedBehavior.php [hunk 5]
```diff
@@ -168,29 +194,30 @@ public function encodeBitmask(Model $Model, $value, $defaultValue = null) {
...
-		}
-
-		foreach ($conditions as $key => $val) {
-			if ($key === $mappedField) {
-				$conditions[$field] = $this->encodeBitmask($Model, $val);
-				if ($field !== $mappedField) {
-					unset($conditions[$mappedField]);
+		foreach ($this->settings[$Model->alias] as $fieldConfig) {
+			$field = $fieldConfig['field'];
+			if (!($mappedField = $fieldConfig['mappedField'])) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: no ia

File: Model/Behavior/BitmaskedBehavior.php [hunk 6]
```diff
@@ -200,80 +227,82 @@ public function encodeBitmaskConditions(Model $Model, $conditions) {
...
 	 */
-	protected function _containsBit(Model $Model, $bits, $contain = true) {
+	protected function _containsBit(Model $Model, $bits, $fieldName, $contain = true) {
 		$bits = (array)$bits;
 		$bitmask = $this->encodeBitmask($Model, $bits);
-
-		$field = $this->settings[$Model->alias]['field'];
 		$contain = $contain ? ' & ? = ?' : ' & ? != ?';
-		return ['(' . $Model->alias . '.' . $field . $contain . ')' => [$bitmask, $bitmask]];
+		return array['(' . $Model->alias . '.' . $fieldName . $contain . ')' => [$bitmask, $bitmask]];
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: We use []

File: Model/Behavior/BitmaskedBehavior.php (hunk 1 above)
Comment (bancer) [inline Model/Behavior/BitmaskedBehavior.php]: Checks if the first element of $config is an array.

File: Model/Behavior/BitmaskedBehavior.php (hunk 4 above)
Comment (bancer) [inline Model/Behavior/BitmaskedBehavior.php]: I'm not sure how that should be changed. That's an old code.

Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: }
if () {
}

Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: but no problem

**Comment by dereuromark** [comment]
We definitively need a few tests for this drastic change of the behavior.

File: Model/Behavior/BitmaskedBehavior.php [hunk 7]
```diff
@@ -60,10 +60,12 @@ class BitmaskedBehavior extends ModelBehavior {
 	public function setup(Model $Model, $config = []) {
 		if (is_array(reset($config))) {
 			foreach ($config as $fieldConfig) {
+				$config += $this->_defaultConfig;
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: shouldnt that be $fieldConfig ?

File: Test/Case/Model/Behavior/BitmaskedBehaviorTest.php [hunk 8]
```diff
@@ -110,6 +110,7 @@ public function testSave() {
 	 * Assert that you can manually trigger "notEmpty" rule with null instead of 0 for "not null" db fields
 	 */
 	public function testSaveWithDefaultValue() {
+		$this->Comment->Behaviors->unload('Tools.Bitmasked');
```
Comment (dereuromark) [inline Test/Case/Model/Behavior/BitmaskedBehaviorTest.php]: unload does not have a Tools. prefix!

File: Model/Behavior/BitmaskedBehavior.php [hunk 9]
```diff
@@ -118,22 +152,38 @@ public function beforeValidate(Model $Model, $options = []) {
...
+	 */
+	protected function _getFieldName(Model $Model) {
+		$firstField = reset($this->settings[$Model->alias]);
+		return $firstField['field'];
+	}
+
+	/**
 	 * @param Model $Model
 	 * @param int $value Bitmask.
+	 * @param string $fieldName field name.
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: |null

File: Model/Behavior/BitmaskedBehavior.php [hunk 10]
```diff
@@ -200,80 +254,88 @@ public function encodeBitmaskConditions(Model $Model, $conditions) {
...
+	public function isBit(Model $Model, $bits, $fieldName = null) {
+		if (empty($fieldName)) {
+			$fieldName = $this->_getFieldName($Model);
+		}
 		$bits = (array)$bits;
 		$bitmask = $this->encodeBitmask($Model, $bits);
-
-		$field = $this->settings[$Model->alias]['field'];
-		return [$Model->alias . '.' . $field => $bitmask];
+		return array($Model->alias . '.' . $fieldName => $bitmask);
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: []

File: Model/Behavior/BitmaskedBehavior.php [hunk 11]
```diff
@@ -200,80 +254,88 @@ public function encodeBitmaskConditions(Model $Model, $conditions) {
...
 
 	/**
 	 * @param Model $Model
 	 * @param mixed $bits (int, array)
+	 * @param string $fieldName field name.
 	 * @param bool $contain
 	 * @return array SQL snippet.
 	 */
-	protected function _containsBit(Model $Model, $bits, $contain = true) {
+	protected function _containsBit(Model $Model, $bits, $fieldName = null, $contain = true) {
```
Comment (dereuromark) [inline Model/Behavior/BitmaskedBehavior.php]: Does $fieldName have to be = null here? It is a protected method.

**Comment by dereuromark** [comment]
Tests seem to pass now.
Would be good to have one more test with the new config and two values like your use case.

For me this looks like a lot of changes compared to attaching a behavior twice, but I am ok with it.

**Comment by dereuromark** [comment]
Awesome :+1:

**Comment by bancer** [comment]
:)

**Pull request merged**